### PR TITLE
Generate one-time keys when we receive new one-time key counts

### PR DESCRIPTION
**Pull request by poljar**
Previously we would generate one-time keys, if needed, whenever we tried to upload them. This code-path is critically missing a `save_account()` call and we would only persist the account once we uploaded the one-time keys.

This patch changes things up to generate one-time keys whenever we receive new one-time key counts from the sync response. This aligns with the way we generate fallback keys and removes the need to introduce a new place where we persist the `Account`.

It's still possible to re-upload the same one-time keys, in the case where the upload process succeeds on the server side but we fail to receive the response.

### CODE REVIEW

**Comment by poljar** [comment]
> I'm not an expert in this area, but this seems reasonable. Does the public API change, requiring users to explicitly generate the keys, under some conditions? (Asking since I'm seeing this change in a few tests now.)

No, the public API did not change. Users would do things the following way:

1. Sync with the server and push the response into the state machine using `OlmMachine::receive_sync_changes()`.
2. Call `OlmMachine::outgoing_requests()` to upload new one-time keys.

No 1. updates the one-time key counts from the sync response. Before this patch we would generate fallback keys at step 1, while we would generate one-time keys at step 2.

After the patch we're generating both one-time keys and fallback keys at step 1.

**Comment by poljar** [comment]
Will test this in a real client before merging, so please don't merge for me.

**Comment by poljar** [comment]
Alright, last fixup commit was needed to retain the exact same upload behavior. Should be good to go.

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/account.rs:572]:
```suggestion
        // This ensures we upload one-time keys along with our device keys right
        // away, rather than waiting for the key counts to be echoed back to us
        // from the server.
```

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/account.rs:576]:
```suggestion
        // It would be nice to do this for the fallback key as well but we can't assume
        // that the server supports fallback keys. Maybe one of these days we
        // will be able to do so.
```

**Review by dkasak — APPROVED** [review]
Functionally this looks good. I suggested some minor wording tweaks to attempt to clarify comments.

**Pull request merged**
